### PR TITLE
[SPARK-47682][SQL] Support cast from variant.

### PR DESCRIPTION
**Pull request by chenhao-db**
### What changes were proposed in this pull request?

This PR allows casting from variant to another type. It has the same semantics as `variant_get` with an empty path.

### Why are the changes needed?

It can bring much convenience if the user can directly cast variant into another type rather than having to use `variant_get` with an empty path.

### Does this PR introduce _any_ user-facing change?

Yes. Casting from variant was previously not allowed but now allowd.

### How was this patch tested?

Unit tests.

### Was this patch authored or co-authored using generative AI tooling?

No.

### CODE REVIEW

**Comment by HyukjinKwon** [comment]
Can you keep the PR template please? https://github.com/apache/spark/blob/master/.github/PULL_REQUEST_TEMPLATE

**Comment by chenhao-db** [comment]
@HyukjinKwon Sure.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Cast.scala [hunk 1]
```diff
@@ -1198,6 +1208,20 @@ case class Cast(
 
     case _ if from == NullType => (c, evPrim, evNull) => code"$evNull = true;"
     case _ if to == from => (c, evPrim, evNull) => code"$evPrim = $c;"
+    case _ if from.isInstanceOf[VariantType] => (c, evPrim, evNull) =>
+      val tmp = ctx.freshVariable("tmp", classOf[Object])
+      val dataTypeArg = ctx.addReferenceObj("dataType", to)
+      val zoneIdArg = ctx.addReferenceObj("zoneId", timeZoneId)
+      val failOnError = evalMode != EvalMode.TRY
+      code"""
+        Object $tmp = org.apache.spark.sql.catalyst.expressions.variant.VariantGet.cast(
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Cast.scala:1217]: nit: use `classOf[T]` to avoid hardcoding class name

**Review by cloud-fan — APPROVED** [review]

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Cast.scala (hunk 1 above)
Comment (chenhao-db) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Cast.scala]: Done.

**Comment by chenhao-db** [comment]
@cloud-fan Could you help review this PR? Thanks!

**Comment by cloud-fan** [comment]
thanks, merging to master!

**Pull request closed without merging**